zebra-network/src/peer_set/set/tests/vectors.rs

```diff
@@ -0,0 +1,63 @@
+use futures::FutureExt;
+use tower::{Service, ServiceExt};
+
+use zebra_chain::parameters::{Network, NetworkUpgrade};
+
+use super::{PeerSetBuilder, PeerVersions};
+use crate::{peer::MinimumPeerVersion, protocol::external::types::Version, Request};
+
+#[test]
+fn peer_set_drop() {
+    //
+    let peer_versions = PeerVersions {
+        peer_versions: vec![Version::min_specified_for_upgrade(
+            Network::Mainnet,
+            NetworkUpgrade::Canopy,
+        )],
+    };
+
+    //
+    let runtime = zebra_test::init_async();
+
+    //
+    let (discovered_peers, _handles) = peer_versions.mock_peer_discovery();
+    let (minimum_peer_version, _best_tip_height) =
+        MinimumPeerVersion::with_mock_chain_tip(Network::Mainnet);
+
+    //
+    runtime.block_on(async move {
+        let (mut peer_set, _peer_set_guard) = PeerSetBuilder::new()
+            .with_discover(discovered_peers)
+            .with_minimum_peer_version(minimum_peer_version.clone())
+            .build();
+
+        // Wait until the peer set is ready
+        let peer_set1 = peer_set.ready().now_or_never();
+
+        //
+        match peer_set1 {
+            Some(ready) => {
+                // make a call that returns a future
+                let fut = ready
+                    .expect("peer set service is always ready")
+                    .call(Request::Peers);
+                // drop the future
+                std::mem::drop(fut);
```

Comment:
This is good.

We'll also need a test that keeps ownership of the future, but doesn't await it, then checks that the request times out. This needs a real `peer::Connection` with a real timeout. (But its channels can be mocked.)

@jvff does `mock_peer_discovery` provide real `peer::Client`s with real `peer::Connection`s?
If it doesn't, we can do that in another ticket or PR.

We'll also need tests that:
- drop the `PeerSet::ready` future
- keeps ownership but does not await the `PeerSet::ready` future

Reply:
> the `MockedClientHandle` associated to the `Client` can be used to receive requests from the peer

I'm not sure if the details matter here, but `Client` gets requests from Zebra, and passes responses back from the peer. (The responses come from the `Connection`, through the `ClientRequest` oneshot channel. But the `Connection` channel is mocked in the tests.)

Reply:
I've added a list of test combinations to the ticket description.

Reply:
I am marking this conversation as we are not going to make all the tests we initially wanted to do. We want to the focus in the last review where tests we already have are improved.